python/src/trezorlib/authentication.py

```diff
@@ -0,0 +1,345 @@
+            "4389bc454f02089d7f08b873005e4c28d432468997871c0bf286fd3861e21e96a"
+        ),
+    ),
+    RootCertificate(
+        "TESTING ENVIRONMENT. DO NOT USE THIS DEVICE",
+        "Trezor Safe 5",
+        True,
+        _pk_p256(
+            "04e48b69cd7962068d3cca3bcc6b1747ef496c1e28b5529e34ad7295215ea161d"
+            "be8fb08ae0479568f9d2cb07630cb3e52f4af0692102da5873559e45e9fa72959"
+        ),
+    ),
+]
+
+
+class DeviceNotAuthentic(Exception):
+    pass
+
+
+class Certificate:
+    def __init__(self, cert_bytes: bytes) -> None:
+        self.cert_bytes = cert_bytes
+        self.cert = x509.load_der_x509_certificate(cert_bytes)
+
+    def __str__(self) -> str:
+        return self.cert.subject.rfc4514_string()
+
+    def public_key_bytes(self) -> bytes:
+        return self.cert.public_key().public_bytes(
+            serialization.Encoding.X962,
+            serialization.PublicFormat.UncompressedPoint,
+        )
+
+    def verify(self, signature: bytes, message: bytes) -> None:
+        cert_pubkey = self.cert.public_key()
+        assert isinstance(cert_pubkey, ec.EllipticCurvePublicKey)
+        cert_pubkey.verify(
+            self.fix_signature(signature),
+            message,
+            ec.ECDSA(hashes.SHA256()),
+        )
+
+    def verify_by(self, pubkey: ec.EllipticCurvePublicKey) -> None:
+        algo_params = self.cert.signature_algorithm_parameters
+        assert isinstance(algo_params, ec.ECDSA)
+        pubkey.verify(
+            self.fix_signature(self.cert.signature),
+            self.cert.tbs_certificate_bytes,
+            algo_params,
+        )
+
+    def is_issued_by(self, issuer: "Certificate", path_len: int) -> bool:
+        """Check if this certificate was issued by an issuer.
+
+        Returns True if:
+         * our `issuer` is the same as issuer's `subject`,
+         * the issuer is a valid CA, that is:
+           - has the cA flag set
+           - has a valid pathLenConstraint
+           - pathLenConstraint does not exceed the current path length.
+         * the issuer's public key signs this certificate.
+
+        Returns False if `issuer` is not permitted to issue this certificate, that is,
+        some of the above conditions are not met.
+
+        Raises InvalidSignature if `issuer` is valid but did not sign this certificate.
+        """
+        if issuer.cert.subject != self.cert.issuer:
+            LOG.error("Certificate %s is not issued by %s.", self, issuer)
+            return False
+
+        try:
+            # check issuer CA flag
+            ext = issuer.cert.extensions.get_extension_for_class(x509.BasicConstraints)
+            if ext.value.ca is not True:
+                LOG.error("Issuer is not a valid CA: %s", issuer)
+                return False
+            if ext.value.path_length is None:
+                LOG.error("Issuer is not a valid CA: %s", issuer)
+                return False
```

Comment:
I think this is actually not an error. BTW, I am not sure right now what we have in our CAs.
> Where pathLenConstraint does not appear, no limit is imposed.

https://datatracker.ietf.org/doc/html/rfc5280#section-4.2.1.9

Reply:
i believe it is fine to enforce that certificates authenticating Trezors _must_ have a path limit set

this is our CA, for the record:
```
Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number: 7356626466875390824 (0x6617fccb67583b68)
        Signature Algorithm: ecdsa-with-SHA256
        Issuer: C = CZ, O = Trezor Company s.r.o., CN = Trezor Manufacturing Root CA
        Validity
            Not Before: Apr 11 15:07:55 2024 GMT
            Not After : Apr  4 15:07:55 2054 GMT
        Subject: C = CZ, O = Trezor Company s.r.o., CN = Trezor Manufacturing CA
        Subject Public Key Info:
            Public Key Algorithm: id-ecPublicKey
                Public-Key: (256 bit)
                pub:
                    04:5b:78:5b:70:38:10:36:3d:ee:ad:c0:fa:a5:a7:
                    38:8d:83:85:f5:e4:f9:17:0a:3c:27:d3:d2:50:6b:
                    f9:09:0d:5b:09:04:4f:ba:d6:b9:06:44:6d:8c:db:
                    75:4b:65:86:9e:89:92:b2:6a:34:4e:01:ec:ff:75:
                    84:4f:e5:a3:d6
                ASN1 OID: prime256v1
                NIST CURVE: P-256
        X509v3 extensions:
            X509v3 Key Usage: critical
                Certificate Sign
            X509v3 Basic Constraints: critical
                CA:TRUE, pathlen:0
            X509v3 Authority Key Identifier: 
                1F:95:3C:3C:B7:93:8F:54:0E:E8:73:52:43:19:E6:06:E0:79:B6:FF
    Signature Algorithm: ecdsa-with-SHA256
    Signature Value:
        30:44:02:20:70:2f:dd:25:73:8a:45:a0:0b:a1:77:7c:6b:78:
        df:fa:be:3f:e5:3b:d3:0c:e8:26:c7:d3:91:c5:b7:0f:ea:83:
        02:20:07:52:71:71:8e:6e:02:dc:eb:9a:79:24:35:fe:16:de:
        5b:0f:08:19:09:d7:da:eb:8e:f7:d4:75:71:a3:d6:ca
```

Reply:
2b33189bdb117d8c6684c69aabd86ca9d755b381

Reply:
ah, except this is not correct because it does not enforce _presence_ of the extensions